tests/python_package_test/test_sklearn.py

```diff
@@ -617,6 +617,104 @@ def test_nan_handle(self):
+            self.assertEqual(len(gbm.evals_result_), len(eval_set_names))
+            for eval_set_name in eval_set_names:
+                self.assertIn(eval_set_name, gbm.evals_result_)
+                self.assertEqual(len(gbm.evals_result_[eval_set_name]), len(metric_names))
+                for metric_name in metric_names:
+                    self.assertIn(metric_name, gbm.evals_result_[eval_set_name])
+
+                    actual = len(gbm.evals_result_[eval_set_name][metric_name])
+                    expected = assumed_iteration + (params_fit['early_stopping_rounds']
+                                                    if eval_set_name != 'training' else 0)
+                    self.assertEqual(actual, expected)
+                    #self.assertEqual(assumed_iteration if eval_set_name != 'training' else 0, gbm.best_iteration_)
+                    self.assertEqual(assumed_iteration, gbm.best_iteration_)
+
+        # training data as eval_set
+        params_fit['eval_set'] = (X_train, y_train)
+        fit_and_check(['training'], ['l2'], 30, False)
+        fit_and_check(['training'], ['l2'], 30, True)
+
+        # # single eval_set
+        iter_1 = 2
+        iter_2 = 8
+        params_fit['eval_set'] = (X_test, y_test)
+        fit_and_check(['valid_0'], ['l2'], iter_1, False)
+        fit_and_check(['valid_0'], ['l2'], iter_1, True)
+
+        params_fit['eval_metric'] = "l2"
+        fit_and_check(['valid_0'], ['l2'], iter_1, False)
+        fit_and_check(['valid_0'], ['l2'], iter_1, True)
+
+        params_fit['eval_metric'] = "l1"
+        fit_and_check(['valid_0'], ['l1', 'l2'], iter_1, False)
+        fit_and_check(['valid_0'], ['l1', 'l2'], iter_2, True)
+
+        params_fit['eval_metric'] = ["l1", "l2"]
+        fit_and_check(['valid_0'], ['l1', 'l2'], iter_1, False)
+        fit_and_check(['valid_0'], ['l1', 'l2'], iter_2, True)
+
+        params_fit['eval_metric'] = ["l2", "l1"]
+        fit_and_check(['valid_0'], ['l1', 'l2'], iter_1, False)
+        fit_and_check(['valid_0'], ['l1', 'l2'], iter_1, True)
+
+        params_fit['eval_metric'] = ["l2", "regression", "mse"]  # test aliases
+        fit_and_check(['valid_0'], ['l2'], iter_1, False)
+        fit_and_check(['valid_0'], ['l2'], iter_1, True)
+
+        # two eval_set
+        X_test1, X_test2, y_test1, y_test2 = train_test_split(X_test, y_test, test_size=0.3, random_state=72)
+        params['learning_rate'] = 0.4
+
+        iter_l2_val0 = 6
+        iter_l2_val1 = 23
+        iter_l1_val0 = 4
+        iter_l1_val1 = 23
+        iter_l2_min =  min([iter_l2_val0, iter_l2_val1])
+        iter_l1_min =  min([iter_l1_val0, iter_l1_val1])
+        best_iter_min = min([iter_l2_val0, iter_l2_val1, iter_l1_val0, iter_l1_val1])
+        params_fit['eval_set'] = [(X_test1, y_test1), (X_test2, y_test2)]
+        fit_and_check(['valid_0', 'valid_1'], ['l2'], iter_l2_val0, True)
+        params_fit['eval_set'] = [(X_test2, y_test2), (X_test1, y_test1)]
```

Comment:
Does it mean that we have different behavior which depends on validation sets order? It's not good, because early stopping mechanism should check all sets and stops for any of them when meet the criterion.
> Early stopping requires at least one set in valid_sets. If there is more than one, it will use all of them except the training data:
   https://github.com/microsoft/LightGBM/blob/master/docs/Python-Intro.rst#early-stopping


https://github.com/microsoft/LightGBM/blob/207bb3ef329b38fc255ef36b1d7ca765924363ea/python-package/lightgbm/engine.py#L96